form/data_mappers.rst

```diff
@@ -0,0 +1,208 @@
+
+            // initialize form field values
+            $forms['red']->setData($data->getRed());
+            $forms['green']->setData($data->getGreen());
+            $forms['blue']->setData($data->getBlue());
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function mapFormsToData($forms, &$data)
+        {
+            /** @var FormInterface[] $forms */
+            $forms = iterator_to_array($forms);
+
+            // as data is passed by reference, overriding it will change it in
+            // the form object as well
+            // beware of type inconsistency, see caution below
+            $data = new Color(
+                $forms['red']->getData(),
+                $forms['green']->getData(),
+                $forms['blue']->getData()
+            );
+        }
+    }
+
+.. caution::
+
+    The data passed to the mapper is *not yet validated*. This means that your
+    objects should allow being created in an invalid state in order to produce
+    user-friendly errors in the form.
+
+Using the Mapper
+----------------
+
+You're ready to use the data mapper for the ``ColorType`` form. Use the
+:method:`Symfony\\Component\\Form\\FormConfigBuilderInterface::setDataMapper`
+method to configure the data mapper::
+
+    // src/App/Form/Type/ColorType.php
+    namespace App\Form\Type;
+
+    use App\Form\DataMapper\ColorMapper;
+    use Symfony\Component\Form\Extension\Core\Type\IntegerType;
+
+    final class ColorType extends AbstractType
+    {
+        /**
+         * {@inheritdoc}
+         */
+        public function buildForm(FormBuilderInterface $builder, array $options)
+        {
+            $builder
+                ->add('red', IntegerType::class, array(
+                    // enforce the strictness of the type to ensure the constructor
+                    // of the Color class doesn't break
+                    'empty_data' => '0',
+                ))
+                ->add('green', IntegerType::class, array(
+                    'empty_data' => '0',
+                ))
+                ->add('blue', IntegerType::class, array(
+                    'empty_data' => '0',
+                ))
+                ->setDataMapper(new ColorMapper())
+            ;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function configureOptions(OptionsResolver $resolver)
+        {
+            $resolver
+                ->setDefaults(array(
+                    // when creating a new color, the initial data should be null
+                    'empty_data' => null,
+                ))
+            ;
+        }
+    }
+
+Cool! When using the ``ColorType`` form, the custom ``ColorMapper`` will create
+a new ``Color`` object now.
+
+.. caution::
+
+    When a form has ``inherit_data`` option set to ``true``, it does not use the data mapper and
+    lets its parent map inner values.
+
+.. tip::
+
+    You can also implement ``DataMapperInterface`` in the ``ColorType`` and add
```

Comment:
[...] implement the `DataMapperInterface` [...]

Reply:
Should we also talk about the possibility to implement the data mapper as an anonymous class?

Reply:
I think in this very case when we have two rather complicated methods to implement, an anonymous class will be harder to read.